### ISSUE TITLE
Support for the Tagging Project

### ISSUE THREAD

**Issue by u-fischer** [pgf-tikz/pgf#1370]
### Brief outline of the proposed feature

I'm currently trying to implement some first support to properly tag tikz pictures and have a few request to tikz/pgf here.

##  Predefine and reserve the keys `alt` and `actualtext` and `artifact`. 

In various cases user will have to provide an alternative text or an actualtext for a picture or declare that the graphic is only decoration (and artifact):

~~~~
\begin{tikzpicture}[alt=A duck]
\duck
\end{tikzpicture}
~~~~

I will provide the needed tagging code, but it would be nice if tikz would predefine the keys for this purpose so that users can add such texts even if tagging is not active (`\includegraphics` knows the `alt` keys already too and I will add `actualtext`).

It would be also nice, if tikz could process these keys earlier,  so that one knows which tagging is needed before the figure is started.   

## Interface for tagging code around nodes.

I must be able to add tagging code around nodes. Currently I'm redefining for this two internal commands of pgf:

~~~~
   \def\pgfsys@begin@text
    {
      \bool_if:NT\l__tag_tikzpicture_usetext_bool
        {\ResumeTagging{\tikzpicture}} 
      \tag_socket_use:n{tikzpicture/textbegin}
    }
   \def\pgfsys@end@text
    {
      \tag_socket_use:n{tikzpicture/textend}
      \bool_if:NT\l__tag_tikzpicture_usetext_bool
        {\SuspendTagging{\tikzpicture} } 
    }
~~~~

This looks more or less ok: as far as I could see the commands are only used for postscript + svg where tagging is not active. But dedicated commands at the same place would be better.

## Getting the BBox

In some case I have to calculate the BBox, these are the absolute coordinates on the page of the bottom/left and top/right corners of the picture. I'm using `remember picture` and then the following code to calculate the coordinates. This works fine (I currently do not care about cases where parts of the picture are outside the bounding box), but here too it would be nice, if there were a public tikz command that one could use:

~~~~
% redefine `\pgfqpoint` locally and then call  `pgf@sys@pdf@mark@pos@pgfid\the\pgf@picture@serial@count`:
   \cs_set:Npn\pgfqpoint##1##2
    {         
      \dim_to_decimal_in_bp:n {##1+ \pgf@picminx}
      \c_space_tl
      \dim_to_decimal_in_bp:n {##2+ \pgf@picminy}
      \c_space_tl
      \dim_to_decimal_in_bp:n {##1+ \pgf@picmaxx}
      \c_space_tl
      \dim_to_decimal_in_bp:n {##2+ \pgf@picmaxx}          
    }   
   \cs_if_exist:cT { pgf@sys@pdf@mark@pos@pgfid\the\pgf@picture@serial@count }
    {
      \__tag_prop_gput:cne
        { g__tag_struct_ \g__tag_struct_stack_current_tl _prop }
        { A } 
        {
           << 
             /O /Layout /BBox~
             [
               \use:c 
                { pgf@sys@pdf@mark@pos@pgfid\the\pgf@picture@serial@count }
             ]
           >>   
        }
     }
~~~~

**Comment by jasperhabicht** [comment]
It would really be nice if the relevant keys could be predefined per default by Ti*k*Z. Currently, one needs to predefine the keys using `\tikzset{actualtext/.code={}, alt/.code={}, artifact/.code={]}` or similar in the package code if one wants to make the package tagging-ready, but it would of course be much better to have these predefined in Ti*k*Z per default.

I think, predefining these keys would not add much to the code base of Ti*k*Z and implementation is probably not too difficult.

**Comment by hmenke** [comment]
@u-fischer In https://github.com/pgf-tikz/pgf/pull/1451 I have now reserved `actualtext`, `alt`, and `artifact`. As for tagging around nodes and the bounding box, I can't really think of a good interface.

For nodes you could use the existing hooks instead of reusing `\pgfsys@begin@text`/`\pgfsys@end@text`
```latex
\tikzset{
  execute at begin node={\ResumeTagging{\tikzpicture}},
  execute at end node={\SuspendTagging{\tikzpicture}},
}
```

The bounding box in page coordinates is extremely tricky, because things like these are totally possible
```latex
\begin{lrbox}{\mybox}
  \begin{tikzpicture}
    % ...
  \end{tikzpicture}
\end{lrbox}
% do whatever transformation and positioning
\usebox{\mybox}
```
and `remember picture` doesn't help there at all. Also `remember picture` on every picture can severely bloat the aux file.

**Comment by u-fischer** [comment]
@hmenke I don't think that it would be sensible to use the public hooks `execute at begin node`  as everyone can overwrite them or add code to them and that would make it be very hard to control tagging. The two commands I used are quite ok, you only need to be aware that they are used for tagging and not remove them or change them. If you think that this could be a problem, you could add some `\pgfsys@begin@text@for@tagging@support`  there.

**Comment by u-fischer** [comment]
> The bounding box in page coordinates is extremely tricky

I know and I do not expect to always get the "correct" bounding box.  But it would be nice to get whatever tikz thinks about the current bounding box and position  in a bit more natural way without having to rely on a tikz internal. 

(Side remark: the current code is slightly different as pgfplots redefines the command and then it is no longer expandable. I now call the command before the structure command:

~~~~
  \use:c { pgf@sys@pdf@mark@pos@pgfid\the\pgf@picture@serial@count }
      \tag_struct_gput:ene{....}
~~~~

**Comment by muzimuzhi** [comment]
> In [#1451](https://github.com/pgf-tikz/pgf/pull/1451) I have now reserved `actualtext`, `alt`, and `artifact`.

#1451 reserved those keys for the `/tikz` default path. Need to do the same for `/pgf` default path?

**Comment by hmenke** [comment]
@muzimuzhi Good call. Do you want to add it?

**Comment by cfr42** [comment]
Shouldn't `/tikz/tagging-setup` be reserved, too? 

The way the code works at present, it is possible for a third-party package to add custom tagging keys by adding them to this family. This ensures the keys are pre-processed along with the `latex-lab` keys in the `init` plug. 

To avoid errors, packages must currently do this conditionally or provide their own compatibility code. But it would be a bit nicer and a bit more efficient if keys could be added to the family regardless, in the same way one can create `tagsupport/...` sockets and have them be automatically part of the tagging code. I.e. the argument here is pretty much the same as the argument for reserving `alt` etc., but the audience is a bit different - packages which need more complex tagging setups e.g. `memoize-ext`, `forest-ext`, `prooftrees`. (Somewhat biased listing, perhaps.)

**Comment by hmenke** [comment]
First of all it should be `/tikz/tagging setup` without hyphen in the TikZ convention. Second I wonder why external packages are messing around in our `/tikz` namespace. Use your own prefix.

**Comment by cfr42** [comment]
> First of all it should be `/tikz/tagging setup` without hyphen in the TikZ convention. Second I wonder why external packages are messing around in our `/tikz` namespace. Use your own prefix.

With respect, that is not possible here without causing complete chaos. 

The tagging support uses 
```latex
\NewTaggingSocketPlug{tikz/picture/init}{default}
  {
    \pgfkeyssavekeyfilterstateto\tikz@temp@tagging@saved@filterstate
    \pgfkeysinstallkeyfilter{/pgf/key~filters/active~families~or~no~family}
      {{/pgf/key~filters/false}{/pgf/key~filters/false}}
    \pgfqkeysactivatesinglefamilyandfilteroptions{/tikz/tagging-setup}{/tikz}{#1}
    \tikz@temp@tagging@saved@filterstate
  }
```
This runs before the picture starts in hook code. There is no difference here between `alt`, `artifact`, `actualtext` and `tagging-setup`. They are all directly under `/tikz`, so I don't think one should be more objectionable than the others. Indeed, the first three all depend on the fourth:
```latex
\AddToHook{package/tikz/after}
  {
    \pgfqkeys{/tikz}
     {
      tagging-setup/.is~family,
      alt/.forward~to=/tikz/tagging-setup/alt,
      alt/.belongs~to~family=/tikz/tagging-setup,
      actualtext/.forward~to=/tikz/tagging-setup/actualtext,
      actualtext/.belongs~to~family=/tikz/tagging-setup,
      artifact/.forward~to=/tikz/tagging-setup/artifact,
      artifact/.belongs~to~family=/tikz/tagging-setup,
      tagging-setup/.code=\pgfqkeys{/tikz/tagging-setup}{#1},
      tagging-setup/.belongs~to~family=/tikz/tagging-setup,
      tagging-setup/text/.code =
       {
         \AssignTaggingSocketPlug{tikz/picture/begin}{text}
         \AssignTaggingSocketPlug{tikz/picture/end}{text}
       },
      tagging-setup/text/.belongs~to~family=/tikz/tagging-setup,
      tagging-setup/alt/.code =
       {
         \tl_if_eq:nnF{#1}{\pgfkeysnovalue}
          { \pdf_purify:nN {#1}\l__tikz_tagging_alt_tl }
         \AssignTaggingSocketPlug{tikz/picture/begin}{alt}
         \AssignTaggingSocketPlug{tikz/picture/end}{alt}
         \def\pgfsys@begin@text{}
         \def\pgfsys@end@text{}
       },
      tagging-setup/alt/.belongs~to~family=/tikz/tagging-setup,
      tagging-setup/tag/.code =
       {
         \AssignStructureRole{graphic}{#1}
         \AssignTaggingSocketPlug{tikz/picture/begin}{alt}
         \AssignTaggingSocketPlug{tikz/picture/end}{alt}
         \def\pgfsys@begin@text{}
         \def\pgfsys@end@text{}
       },
      tagging-setup/tag/.belongs~to~family=/tikz/tagging-setup,
      tagging-setup/actualtext/.code =
       {
          \tl_if_eq:nnF{#1}{\pgfkeysnovalue}
           { \pdf_purify:nN {#1} \l__tikz_tagging_actualtext_tl }
          \AssignTaggingSocketPlug{tikz/picture/begin}{actualtext}
          \AssignTaggingSocketPlug{tikz/picture/end}{actualtext}
          \def\pgfsys@begin@text{}
          \def\pgfsys@end@text{}
       },
      tagging-setup/actualtext/.belongs~to~family=/tikz/tagging-setup,
      tagging-setup/artifact/.code =
       {
         \AssignTaggingSocketPlug{tikz/picture/begin}{artifact}
         \AssignTaggingSocketPlug{tikz/picture/end}{artifact}
         \def\pgfsys@begin@text{}
         \def\pgfsys@end@text{}
       },
      tagging-setup/artifact/.belongs~to~family=/tikz/tagging-setup,
      tagging-setup/off/.code =
        {
          \AssignTaggingSocketPlug{tikz/picture/begin}{noop}
          \AssignTaggingSocketPlug{tikz/picture/end}{noop}
          \def\pgfsys@begin@text{}
          \def\pgfsys@end@text{}
        },
      tagging-setup/off/.belongs~to~family=/tikz/tagging-setup,
     }
 }
```
This is required because the tagging code must be set up before the picture starts, so we have to pre-process a subset of the keys the user passes to `tikzpicture` or `pgfpicture` etc. We do not want to process all of the keys there. So the obvious way to do this is to filter out the keys we want. These are the default keys, but we should also process any custom tagging keys the user or other packages setup. 

For the very vast majority of cases, it should not be necessary to add additional sockets or write custom replacements for `init`. All that's necessary is adding custom keys to the `/tikz/tagging-support` family. This is equivalent, IMHO, to creating sockets in the `tagsupport` namespace, which third party code is supposed to do if the sockets are used for tagging. This way commands to temporarily suspend or resume tagging work as expected. IMHO, the `pgfkeys` interface should reflect the same general approach i.e. it should provide supported interfaces for authors and packages to customise tagging. 

About the hyphen. I originally used an internal name for the family (`tagging@setup`?) and provided a public handler so packages could declare custom keys, but @u-fischer  wanted to avoid the handler and used the sub-tree name also for the family, which also works perfectly well. If you do not like the name, you can, of course, use a space, but this would need to be coordinated with changes in `latex-lab-testphase-tikz.sty`, which currently uses the hyphen. 

But if you do not like the general approach of using a family to filter etc., you should blame me rather than @u-fischer.

**Comment by hmenke** [comment]
No, this particular name is not at all required for tagging support. You could equally use something like `/latex-lab/tikz/tagging-support` (also including as many hyphens as you please). Using `/tikz/tagging-support` suggest to a user that this is something provided by TikZ which it is not.

**Comment by cfr42** [comment]
> No, this particular name is not at all required for tagging support. You could equally use something like `/latex-lab/tikz/tagging-support` (also including as many hyphens as you please). Using `/tikz/tagging-support` suggest to a user that this is something provided by TikZ which it is not.

But that is, after all, the plan, is it not? That TikZ will not perpetually need first-aid patches to work with tagging. Otherwise, why add `/tikz/alt` etc? Those could equally use some similar namespace.

**Comment by cfr42** [comment]
Obviously, it is true that no specific name is required for tagging support. But that's true of every key provided by the first-aid.

**Comment by cfr42** [comment]
To put this differently, why are you OK adding only part of the `pgfkeys` support here? Setting aside the question of what any particular key should be called, a user may do
```latex
\tikzset{alt=Nice picture}
```
or 
```latex
\tikzset{tagging-setup/text}
```
The latter is the default, so not normally needed, but it is required if the default is changed for part of the document and one then wishes to restore the default plug.

And, yes, `latex-lab` could have used its own `pgfkeys` prefix here. But it didn't. 

I could understand your saying: I don't like this, `latex-lab` shouldn't have done it and I'm not going to support it. If people want tagging, they can get it elsewhere.

What I don't understand is your saying: I'm happy to support *some* keys `latex-lab` wants for tagging, but I draw the line at supporting 4. 3 is my absolute limit.

**Comment by u-fischer** [comment]
> why external packages are messing around in our /tikz namespace.

Well the three keys `alt`, `artifact` and `actualtext` should imho always work. I do not care in which name space they are as long as something like `\tikz[alt=a circle]{...}` works as intended.  

I also think the more generic tagging setup key is needed, but I do no care much about the name.  If you have some other suggestion we can adapt.

Please note that the keys for tagging must be processed rather early, so I patch into `\tikz@picture`, `\pgfpicture` and `endpgfpicture`.

**Comment by cfr42** [comment]
> > why external packages are messing around in our /tikz namespace.
> 
> Well the three keys `alt`, `artifact` and `actualtext` should imho always work. I do not care in which name space they are as long as something like `\tikz[alt=a circle]{...}` works as intended.

I am not sure there is a nice way to do this without having them in the `tikz`/`pgf` namespace. It isn't a problem as far as the tagging code in `latex-lab` is concerned since that can obviously set the path to whatever before filtering. But the keys will error when `tikz`/`pgf` processes them. 

In any case, this is more confusing for users. Normally, you'd expect the same keys to work in `\tikzset{}` or `\pgfkeys{}` as in the optional argument to the environment/command. 
```latex
\tikzset{actualtext=something}
\tikz{...}
\tikzset{tagging-setup=artifact}
\tikz{...}
\tikzset{tagging-setup/text}
\tikz{...}
```
It gets a bit weird if these give errors while
```latex
\tikz[actualtext=something]{...}
\tikz[tagging-setup=artifact]{...}
\tikz[tagging-setup/text]{...}
```
work.

> I also think the more generic tagging setup key is needed, but I do no care much about the name. If you have some other suggestion we can adapt.

`tagging`? Avoids both spaces and hyphens ...
